go/ct/gen/transient.go

```diff
@@ -0,0 +1,149 @@
+//
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by the GNU Lesser General Public Licence v3
+//
+
+package gen
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"golang.org/x/exp/maps"
+	"pgregory.net/rand"
+
+	"github.com/Fantom-foundation/Tosca/go/ct/common"
+	"github.com/Fantom-foundation/Tosca/go/ct/st"
+)
+
+type TransientGenerator struct {
+	set           map[Variable]bool
+	unsatisfiable bool
+}
+
+func NewTransientGenerator() *TransientGenerator {
+	return &TransientGenerator{set: map[Variable]bool{}}
+}
+
+func (t *TransientGenerator) BindSet(key Variable) {
+	if value, exits := t.set[key]; exits {
+		if !value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.set[key] = true
+	}
+}
+
+func (t *TransientGenerator) BindNotSet(key Variable) {
+	if value, exits := t.set[key]; exits {
+		if value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.set[key] = false
+	}
+}
+
+func (t *TransientGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.Transient, error) {
+	if t.unsatisfiable {
+		return nil, fmt.Errorf("%w, conflicting set/unset constraints", ErrUnsatisfiable)
+	}
+
+	keysInUse := map[common.U256]bool{}
+	for variable := range t.set {
+		if pos, isBound := assignment[variable]; isBound {
+			keysInUse[pos] = true
+		}
+	}
+
+	getUnusedKey := func() common.U256 {
+		for {
+			key := common.RandU256(rnd)
+			if _, isPresent := keysInUse[key]; !isPresent {
+				keysInUse[key] = true
+				return key
+			}
+		}
+	}
+
+	transient := &st.Transient{}
+
+	// Process set/unset constraints.
+	for variable, set := range t.set {
+
+		// Check different assignments with the same key if they conflict
+		for other, otherSet := range t.set {
+			if assignment[variable] == assignment[other] && set != otherSet {
+				return nil, fmt.Errorf("%w, conflicting set/unset constraints", ErrUnsatisfiable)
+			}
+		}
+
+		var key common.U256
+		if pos, isBound := assignment[variable]; isBound {
+			key = pos
+		} else {
+			key = getUnusedKey()
+		}
+
+		if set {
+			transient.SetStorage(key, common.RandU256(rnd))
+		} else {
+			transient.DeleteStorage(key)
```

Comment:
nit: I might be wrong, but this should not be necessary. The `key` to be deleted should at this point never be set before. If it would have, another constraint that got fulfilled by adding the key gets violated by removing it again.

Removing this simplifies the generator and may allow you to remove the `DeleteStorage` function.

Reply:
I reworked the whole PR and worked on naming, I these changes should make it much more readable.